File: platform/android/MapLibreAndroid/src/main/java/org/maplibre/android/maps/NativeMap.java
package org.maplibre.android.maps;

import android.graphics.Bitmap;
import android.graphics.PointF;
import android.graphics.RectF;

import androidx.annotation.IntRange;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.maplibre.geojson.Feature;
import org.maplibre.geojson.Geometry;
import org.maplibre.android.annotations.Marker;
import org.maplibre.android.annotations.Polygon;
import org.maplibre.android.annotations.Polyline;
import org.maplibre.android.camera.CameraPosition;
import org.maplibre.android.geometry.LatLng;
import org.maplibre.android.geometry.LatLngBounds;
import org.maplibre.android.geometry.ProjectedMeters;
import org.maplibre.android.style.expressions.Expression;
import org.maplibre.android.style.layers.Layer;
import org.maplibre.android.style.layers.TransitionOptions;
import org.maplibre.android.style.light.Light;
import org.maplibre.android.style.sources.Source;

import java.util.List;

interface NativeMap {

  //
  // Lifecycle API
  //

  void resizeView(int width, int height);

  void onLowMemory();

  void destroy();

  boolean isDestroyed();

  //
  // Camera API
  //

  void jumpTo(@NonNull LatLng center, double zoom, double pitch, double bearing, double[] padding);

  void easeTo(@NonNull LatLng center, double zoom, double bearing, double pitch, double[] padding, long duration,
              boolean easingInterpolator);

  void flyTo(@NonNull LatLng center, double zoom, double bearing, double pitch, double[] padding, long duration);

  void moveBy(double deltaX, double deltaY, long duration);

  @NonNull
  CameraPosition getCameraPosition();

  CameraPosition getCameraForLatLngBounds(@NonNull LatLngBounds bounds, int[] padding, double bearing, double pitch);

  CameraPosition getCameraForGeometry(@NonNull Geometry geometry, int[] padding, double bearing, double pitch);

  void resetPosition();

  void setLatLng(@NonNull LatLng latLng, long duration);

  LatLng getLatLng();

  void setLatLngBounds(@Nullable LatLngBounds latLngBounds);

  void setVisibleCoordinateBounds(@NonNull LatLng[] coordinates, @NonNull RectF padding,
                                  double direction, long duration);

  void setPitch(double pitch, long duration);

  double getPitch();

  void setZoom(double zoom, @NonNull PointF focalPoint, long duration);

  double getZoom();

  void setMinZoom(double zoom);

  double getMinZoom();

  void setMaxZoom(double zoom);

  double getMaxZoom();

  void setMinPitch(double pitch);

  double getMinPitch();

  void setMaxPitch(double pitch);

  double getMaxPitch();

  void resetZoom();

  void rotateBy(double sx, double sy, double ex, double ey, long duration);

  void setBearing(double degrees, long duration);

  void setBearing(double degrees, double fx, double fy, long duration);

  double getBearing();

  void resetNorth();

  void cancelTransitions();

  //
  // Style API
  //

  void setStyleUri(String url);

  @NonNull
  String getStyleUri();

  void setStyleJson(String newStyleJson);

  @NonNull
  String getStyleJson();

  boolean isFullyLoaded();

  void addLayer(@NonNull Layer layer);

  void addLayerBelow(@NonNull Layer layer, @NonNull String below);

  void addLayerAbove(@NonNull Layer layer, @NonNull String above);

  void addLayerAt(@NonNull Layer layer, @IntRange(from = 0) int index);

  @NonNull
  List<Layer> getLayers();

  Layer getLayer(String layerId);

  boolean removeLayer(@NonNull String layerId);

  boolean removeLayer(@NonNull Layer layer);

  boolean removeLayerAt(@IntRange(from = 0) int index);

  void addSource(@NonNull Source source);

  @NonNull
  List<Source> getSources();

  Source getSource(@NonNull String sourceId);

  boolean removeSource(@NonNull String sourceId);

  boolean removeSource(@NonNull Source source);

  void setTransitionOptions(@NonNull TransitionOptions transitionOptions);

  @NonNull
  TransitionOptions getTransitionOptions();

  void addImages(Image[] images);

  Bitmap getImage(String name);

  void removeImage(String name);

  Light getLight();

  //
  // Content padding API
  //

  void setContentPadding(double[] padding);

  double[] getContentPadding();

  //
  // Query API
  //

  @NonNull
  List<Feature> queryRenderedFeatures(@NonNull PointF coordinates,
                                      @Nullable String[] layerIds,
                                      @Nullable Expression filter);

  @NonNull
  List<Feature> queryRenderedFeatures(@NonNull RectF coordinates,
                                      @Nullable String[] layerIds,
                                      @Nullable Expression filter);

  //
  // Projection API
  //

  double getMetersPerPixelAtLatitude(double lat);

  ProjectedMeters projectedMetersForLatLng(@NonNull LatLng latLng);

  LatLng latLngForProjectedMeters(@NonNull ProjectedMeters projectedMeters);

  @NonNull
  PointF pixelForLatLng(@NonNull LatLng latLng);

  void pixelsForLatLngs(@NonNull double[] input, @NonNull double[] output);

  void getVisibleCoordinateBounds(@NonNull double[] output);

  LatLng latLngForPixel(@NonNull PointF pixel);

  void latLngsForPixels(@NonNull double[] input, @NonNull double[] output);

  //
  // Utils API
  //

  void setOnFpsChangedListener(@NonNull MapLibreMap.OnFpsChangedListener listener);

  void setDebug(boolean debug);

  boolean getDebug();

  void setReachability(boolean status);

  void setApiBaseUrl(String baseUrl);

  void setPrefetchTiles(boolean enable);

  boolean getPrefetchTiles();

  void setPrefetchZoomDelta(@IntRange(from = 0) int delta);

  @IntRange(from = 0)
  int getPrefetchZoomDelta();

  void setGestureInProgress(boolean inProgress);

  float getPixelRatio();

  void triggerRepaint();

  void setSwapBehaviorFlush(boolean flush);

  //
  // Deprecated Annotations API
  //

  long addMarker(Marker marker);

  @NonNull
  long[] addMarkers(@NonNull List<Marker> markers);

  long addPolyline(Polyline polyline);

  @NonNull
  long[] addPolylines(@NonNull List<Polyline> polylines);

  long addPolygon(Polygon polygon);

  @NonNull
  long[] addPolygons(@NonNull List<Polygon> polygons);

  void updateMarker(@NonNull Marker marker);

  void updatePolygon(@NonNull Polygon polygon);

  void updatePolyline(@NonNull Polyline polyline);

  void removeAnnotation(long id);

  void removeAnnotations(long[] ids);

  double getTopOffsetPixelsForAnnotationSymbol(String symbolName);

  void addAnnotationIcon(String symbol, int width, int height, float scale, byte[] pixels);

  void removeAnnotationIcon(String symbol);

  @NonNull
  long[] queryPointAnnotations(RectF rectF);

  @NonNull
  long[] queryShapeAnnotations(RectF rectF);

  @NonNull
  RectF getDensityDependantRectangle(RectF rectangle);

  long getNativePtr();

  void addSnapshotCallback(@NonNull MapLibreMap.SnapshotReadyCallback callback);
}

File: platform/android/MapLibreAndroid/src/main/java/org/maplibre/android/maps/NativeMap.kt
package org.maplibre.android.maps

import android.graphics.Bitmap
import android.graphics.PointF
import android.graphics.RectF
import androidx.annotation.IntRange
import com.google.gson.JsonObject
import org.maplibre.android.annotations.Marker
import org.maplibre.android.annotations.Polygon
import org.maplibre.android.annotations.Polyline
import org.maplibre.android.camera.CameraPosition
import org.maplibre.android.geometry.LatLng
import org.maplibre.android.geometry.LatLngBounds
import org.maplibre.android.geometry.ProjectedMeters
import org.maplibre.android.style.expressions.Expression
import org.maplibre.android.style.layers.Layer
import org.maplibre.android.style.layers.TransitionOptions
import org.maplibre.android.style.light.Light
import org.maplibre.android.style.sources.Source
import org.maplibre.geojson.Feature
import org.maplibre.geojson.Geometry

@Suppress("TooManyFunctions")
internal interface NativeMap {
    //
    // Lifecycle API
    //

    fun resizeView(
        width: Int,
        height: Int,
    )

    fun onLowMemory()

    fun destroy()

    val isDestroyed: Boolean

    //
    // Camera API
    //

    fun jumpTo(
        center: LatLng,
        zoom: Double,
        pitch: Double,
        bearing: Double,
        padding: DoubleArray?,
    )

    @Suppress("LongParameterList")
    fun easeTo(
        center: LatLng,
        zoom: Double,
        bearing: Double,
        pitch: Double,
        padding: DoubleArray?,
        duration: Long,
        easingInterpolator: Boolean,
    )

    @Suppress("LongParameterList")
    fun flyTo(
        center: LatLng,
        zoom: Double,
        bearing: Double,
        pitch: Double,
        padding: DoubleArray?,
        duration: Long,
    )

    fun moveBy(
        deltaX: Double,
        deltaY: Double,
        duration: Long,
    )

    val cameraPosition: CameraPosition

    // Note for implementors: the ordering of the padding is left, top, right, bottom
    fun getCameraForLatLngBounds(
        bounds: LatLngBounds,
        padding: IntArray,
        bearing: Double,
        pitch: Double,
    ): CameraPosition?

    fun getCameraForGeometry(
        geometry: Geometry,
        padding: IntArray,
        bearing: Double,
        pitch: Double,
    ): CameraPosition?

    fun resetPosition()

    fun setLatLng(
        latLng: LatLng,
        duration: Long,
    )

    val latLng: LatLng

    fun setLatLngBounds(latLngBounds: LatLngBounds?)

    fun setVisibleCoordinateBounds(
        coordinates: Array<LatLng>,
        padding: RectF,
        direction: Double,
        duration: Long,
    )

    fun setPitch(
        pitch: Double,
        duration: Long,
    )

    val pitch: Double

    fun setZoom(
        zoom: Double,
        focalPoint: PointF,
        duration: Long,
    )

    val zoom: Double

    var minZoom: Double

    var maxZoom: Double

    var minPitch: Double

    var maxPitch: Double

    fun resetZoom()

    fun rotateBy(
        sx: Double,
        sy: Double,
        ex: Double,
        ey: Double,
        duration: Long,
    )

    fun setBearing(
        degrees: Double,
        duration: Long,
    )

    fun setBearing(
        degrees: Double,
        fx: Double,
        fy: Double,
        duration: Long,
    )

    val bearing: Double

    fun resetNorth()

    fun cancelTransitions()

    //
    // Style API
    //

    var styleUri: String

    var styleJson: String

    fun isFullyLoaded(): Boolean

    fun addLayer(layer: Layer)

    fun addLayerBelow(
        layer: Layer,
        below: String,
    )

    fun addLayerAbove(
        layer: Layer,
        above: String,
    )

    fun addLayerAt(
        layer: Layer,
        @IntRange(from = 0) index: Int,
    )

    fun getLayers(): List<Layer>

    fun getLayer(layerId: String): Layer?

    fun removeLayer(layerId: String): Boolean

    fun removeLayer(layer: Layer): Boolean

    fun removeLayerAt(
        @IntRange(from = 0) index: Int,
    ): Boolean

    fun addSource(source: Source)

    fun getSources(): List<Source>

    fun getSource(sourceId: String): Source?

    fun removeSource(sourceId: String): Boolean

    fun removeSource(source: Source): Boolean

    var transitionOptions: TransitionOptions

    fun addImages(images: Array<Image>)

    fun getImage(name: String): Bitmap?

    fun removeImage(name: String)

    fun getLight(): Light?

    //
    // Content padding API
    //

    var contentPadding: DoubleArray?

    //
    // Query API
    //

    fun queryRenderedFeatures(
        coordinates: PointF,
        layerIds: Array<out String>?,
        filter: Expression?,
    ): List<Feature>

    fun queryRenderedFeatures(
        coordinates: RectF,
        layerIds: Array<out String>?,
        filter: Expression?,
    ): List<Feature>

    fun setFeatureState(
        sourceId: String,
        sourceLayerId: String?,
        featureId: String,
        state: JsonObject,
    )

    fun getFeatureState(
        sourceId: String,
        sourceLayerId: String?,
        featureId: String,
    ): JsonObject?

    fun removeFeatureState(
        sourceId: String,
        sourceLayerId: String?,
        featureId: String?,
        stateKey: String?,
    )

    //
    // Projection API
    //

    fun getMetersPerPixelAtLatitude(lat: Double): Double

    fun projectedMetersForLatLng(latLng: LatLng): ProjectedMeters

    fun latLngForProjectedMeters(projectedMeters: ProjectedMeters): LatLng

    fun pixelForLatLng(latLng: LatLng): PointF

    fun pixelsForLatLngs(
        input: DoubleArray,
        output: DoubleArray,
    )

    fun getVisibleCoordinateBounds(output: DoubleArray)

    fun latLngForPixel(pixel: PointF): LatLng

    fun latLngsForPixels(
        input: DoubleArray,
        output: DoubleArray,
    )

    //
    // Utils API
    //

    fun setOnFpsChangedListener(listener: MapLibreMap.OnFpsChangedListener?)

    fun setDebug(debug: Boolean)

    fun getDebug(): Boolean

    fun getActionJournalLogFiles(): Array<String>

    fun getActionJournalLog(): Array<String>

    fun clearActionJournalLog()

    fun setReachability(status: Boolean)

    fun setApiBaseUrl(baseUrl: String)

    var prefetchTiles: Boolean

    @setparam:IntRange(from = 0)
    @get:IntRange(from = 0)
    var prefetchZoomDelta: Int

    fun setTileCacheEnabled(enabled: Boolean)

    fun getTileCacheEnabled(): Boolean

    fun setTileLodMinRadius(radius: Double)

    fun getTileLodMinRadius(): Double

    fun setTileLodScale(scale: Double)

    fun getTileLodScale(): Double

    fun setTileLodPitchThreshold(threshold: Double)

    fun getTileLodPitchThreshold(): Double

    fun setTileLodZoomShift(shift: Double)

    fun getTileLodZoomShift(): Double

    fun setGestureInProgress(inProgress: Boolean)

    fun getPixelRatio(): Float

    fun triggerRepaint()

    fun isRenderingStatsViewEnabled(): Boolean

    fun enableRenderingStatsView(value: Boolean)

    fun setFrustumOffset(offset: RectF)

    fun setSwapBehaviorFlush(flush: Boolean)

    //
    // Deprecated Annotations API
    //

    fun addMarker(marker: Marker): Long

    fun addMarkers(markers: List<Marker>): LongArray

    fun addPolyline(polyline: Polyline): Long

    fun addPolylines(polylines: List<Polyline>): LongArray

    fun addPolygon(polygon: Polygon): Long

    fun addPolygons(polygons: List<Polygon>): LongArray

    fun updateMarker(marker: Marker)

    fun updatePolygon(polygon: Polygon)

    fun updatePolyline(polyline: Polyline)

    fun removeAnnotation(id: Long)

    fun removeAnnotations(ids: LongArray)

    fun getTopOffsetPixelsForAnnotationSymbol(symbolName: String): Double

    fun addAnnotationIcon(
        symbol: String,
        width: Int,
        height: Int,
        scale: Float,
        pixels: ByteArray,
    )

    fun removeAnnotationIcon(symbol: String)

    fun queryPointAnnotations(rectF: RectF): LongArray

    fun queryShapeAnnotations(rectF: RectF): LongArray

    fun getDensityDependantRectangle(rectangle: RectF): RectF

    val nativePtr: Long

    fun addSnapshotCallback(callback: MapLibreMap.SnapshotReadyCallback)
}
